File: my-folder/0948-sort-an-array/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArray = function(nums) {
    let tempArr = Array(nums.length).fill(0)

    let merge = (left, mid, right) => {
        let length1 = mid - left + 1
        let length2 = right - mid

        let start1 = left
        let start2 = mid + 1

        for (let i = 0; i < length1; i++) {
            tempArr[start1 + i] = nums[start1 + i]
        }
        for (let j = 0; j < length2; j++) {
            tempArr[start2 + j] = nums[start2 + j]
        }

        let i = 0, j = 0, k = left
        while (i < length1 && j < length2) {
            if (tempArr[start1 + i] < tempArr[start2 + j]) {
                nums[k] = tempArr[start1 + i]
                i += 1
            } else {
                nums[k] = tempArr[start2 + j]
                j += 1
            }
            k += 1
        }

        while (i < length1) {
            nums[k] = tempArr[start1 + i]
            i += 1
            k += 1
        }
        while (j < length2) {
            nums[k] = tempArr[start2 + j]
            j += 1
            k += 1
        }
    }

    let mergeSort = (left, right) => {
        if (left >= right) return

        let mid = Math.floor((left + right) / 2)

        mergeSort(left, mid)
        mergeSort(mid + 1, right)

        merge(left, mid, right)
    }

    mergeSort(0, nums.length - 1)
    return nums
};

File: my-folder/0948-sort-an-array/solution.cs
public class Solution {
    public int[] SortArray(int[] nums) {
        int[] tool = new int[nums.Length];
        MergeSort(nums, tool, 0, nums.Length - 1);

        return nums;
    }

    // locate, divide, then start to Merge it
    private void MergeSort(int[] array, int[] tool, int start, int end) {
        if (start < end) {
            int mid = (start + end) / 2;
            MergeSort(array, tool, start, mid);
            MergeSort(array, tool, mid + 1, end);

            Merge(array, tool, start, mid, end);
        }
    }

    private void Merge(int[] array, int[] tool, int start, int mid, int end) {
        int head1 = start;
        int head2 = mid + 1;
        int head3 = start;
        
        while (head1 <= mid && head2 <= end) {
            if (array[head1] <= array[head2]) {
                tool[head3++] = array[head1++];
            }

            else {
                tool[head3++] = array[head2++];
            }
        }

        while (head1 <= mid) {
            tool[head3++] = array[head1++];
        }

        while (head2 <= end) {
            tool[head3++] = array[head2++];
        }

        for (int i = start; i <= end; i++) {
            array[i] = tool[i];
        }
    }
}
